### ISSUE TITLE
Compile safety throws errors for Android Project

### ISSUE THREAD

**Issue by sarn0ld** [InsertKoinIO/koin-annotations#96]
**Describe the bug**
I tried the [new ksp flag "KOIN_CONFIG_CHECK"](https://insert-koin.io/docs/reference/koin-annotations/start#compile-safety---check-your-koin-config-at-compile-time-since-130) in an Android project.

Now I get compilation errors for Android specific classes:
```
e: [ksp] --> Missing Definition type 'android.content.Context' for '***'. Fix your configuration to define type 'Context'.
e: [ksp] --> Missing Definition type 'androidx.lifecycle.SavedStateHandle' for '***'. Fix your configuration to define type 'SavedStateHandle'.
```

**To Reproduce**
1. Add the ksp argument KOIN_CONFIG_CHECK to an Android project that uses `Context` and `SavedStateHandle` in it's dependency tree
2. Click on compile
3. See Compilation errors in the build output

**Expected behavior**
No compile errors

**Koin module and version:**
```
    implementation "io.insert-koin:koin-androidx-compose:3.5.0"
    implementation "io.insert-koin:koin-annotations:1.3.0"
    testImplementation "io.insert-koin:koin-test:3.5.0"
    testImplementation "io.insert-koin:koin-test-junit5:3.5.0"
    testImplementation 'org.jetbrains.kotlinx:kotlinx-coroutines-test:1.7.1'
    ksp 'io.insert-koin:koin-ksp-compiler:1.3.0'
```

**Comment by alexandrucaraus** [comment]
I have the same problem when using annotations and have the KOIN_CONFIG_CHECK enabled.

**Comment by ghasemdev** [comment]
I create an PR https://github.com/InsertKoinIO/koin-annotations/pull/101

**Comment by ghasemdev** [comment]
> ```
> e: [ksp] --> Missing Definition type 'android.app.Application' for '***'. Fix your configuration to define type 'Application'.
> e: [ksp] --> Missing Definition type 'kotlinx.coroutines.CoroutineDispatcher' for '***'. Fix your configuration to define type 'CoroutineDispatcher'.
> e: [ksp] --> Missing Definition type 'android.content.Context' for '***'. Fix your configuration to define type 'Context'.
> e: [ksp] --> Missing Definition type 'kotlin.String' for '***'. Fix your configuration to define type 'String'.
> e: [ksp] --> Missing Definition type 'android.content.SharedPreferences' for '***'. Fix your configuration to define type 'SharedPreferences'.
> e: [ksp] --> Missing Definition type 'androidx.work.WorkManager' for '***'. Fix your configuration to define type 'WorkManager'.
> e: [ksp] --> Missing Definition type 'androidx.work.OneTimeWorkRequest' for '***'. Fix your configuration to define type 'OneTimeWorkRequest'.
> ```
> 
> I have a few more issues with this ticket.
> 
> cc: @ghasemdev

I am also experiencing this issue in the sample project I’m working on when using the regular provision of koin.

```
e: [ksp] --> Missing Definition type 'com.parsuomash.sdk.di.Token' for 'com.parsuomash.sdk.domain.usecase.Koo'. Fix your configuration to define type 'Token'.
e: [ksp] --> Missing Definition type 'kotlinx.coroutines.CoroutineDispatcher' for 'com.parsuomash.sdk.domain.usecase.UseCase'. Fix your configuration to define type 'CoroutineDispatcher'.
e: [ksp] --> Missing Definition type 'android.content.SharedPreferences' for 'com.parsuomash.sdk.domain.usecase.UseCase'. Fix your configuration to define type 'SharedPreferences'.
e: [ksp] --> Missing Definition type 'android.app.Application' for 'com.parsuomash.sdk.di.Session'. Fix your configuration to define type 'Application'.
e: [ksp] --> Missing Definition type 'android.content.Context' for 'com.parsuomash.sdk.di.Session'. Fix your configuration to define type 'Context'.
e: [ksp] --> Missing Definition type 'androidx.lifecycle.SavedStateHandle' for 'com.parsuomash.sdk.SDKViewModel'. Fix your configuration to define type 'SavedStateHandle'.
e: [ksp] --> Missing Definition type 'com.parsuomash.sdk.SDK' for 'com.parsuomash.sdk.SDKViewModel'. Fix your configuration to define type 'SDK'.
```


```kotlin
internal val tokenModule = module {
  factory {
    Token("dkjgfiuebvldvyl")
  }
}

@JvmInline
value class Token(val value: String)
```

**Comment by ghasemdev** [comment]
Can you send the codes for providing the WorkManager and where you used it?

@I-sty

**Comment by ghasemdev** [comment]
> > ```
> > e: [ksp] --> Missing Definition type 'android.app.Application' for '***'. Fix your configuration to define type 'Application'.
> > e: [ksp] --> Missing Definition type 'kotlinx.coroutines.CoroutineDispatcher' for '***'. Fix your configuration to define type 'CoroutineDispatcher'.
> > e: [ksp] --> Missing Definition type 'android.content.Context' for '***'. Fix your configuration to define type 'Context'.
> > e: [ksp] --> Missing Definition type 'kotlin.String' for '***'. Fix your configuration to define type 'String'.
> > e: [ksp] --> Missing Definition type 'android.content.SharedPreferences' for '***'. Fix your configuration to define type 'SharedPreferences'.
> > e: [ksp] --> Missing Definition type 'androidx.work.WorkManager' for '***'. Fix your configuration to define type 'WorkManager'.
> > e: [ksp] --> Missing Definition type 'androidx.work.OneTimeWorkRequest' for '***'. Fix your configuration to define type 'OneTimeWorkRequest'.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I have a few more issues with this ticket.
> > cc: @ghasemdev
> 
> I am also experiencing this issue in the sample project I’m working on when using the regular provision of koin.
> 
> ```
> e: [ksp] --> Missing Definition type 'com.parsuomash.sdk.di.Token' for 'com.parsuomash.sdk.domain.usecase.Koo'. Fix your configuration to define type 'Token'.
> e: [ksp] --> Missing Definition type 'kotlinx.coroutines.CoroutineDispatcher' for 'com.parsuomash.sdk.domain.usecase.UseCase'. Fix your configuration to define type 'CoroutineDispatcher'.
> e: [ksp] --> Missing Definition type 'android.content.SharedPreferences' for 'com.parsuomash.sdk.domain.usecase.UseCase'. Fix your configuration to define type 'SharedPreferences'.
> e: [ksp] --> Missing Definition type 'android.app.Application' for 'com.parsuomash.sdk.di.Session'. Fix your configuration to define type 'Application'.
> e: [ksp] --> Missing Definition type 'android.content.Context' for 'com.parsuomash.sdk.di.Session'. Fix your configuration to define type 'Context'.
> e: [ksp] --> Missing Definition type 'androidx.lifecycle.SavedStateHandle' for 'com.parsuomash.sdk.SDKViewModel'. Fix your configuration to define type 'SavedStateHandle'.
> e: [ksp] --> Missing Definition type 'com.parsuomash.sdk.SDK' for 'com.parsuomash.sdk.SDKViewModel'. Fix your configuration to define type 'SDK'.
> ```
> 
> ```kotlin
> internal val tokenModule = module {
>   factory {
>     Token("dkjgfiuebvldvyl")
>   }
> }
> 
> @JvmInline
> value class Token(val value: String)
> 
> 
> internal val coroutinesModule = module {
>   single {
>     Dispatchers.IO
>   } withOptions {
>     named("IO")
>     createdAtStart()
>   }
> 
>   single {
>     Dispatchers.Default
>   } withOptions {
>     named("Default")
>     createdAtStart()
>   }
> 
>   single {
>     Dispatchers.Main
>   } withOptions {
>     named("Main")
>     createdAtStart()
>   }
> }
> 
> internal val sharedPrefModule = module {
>   single {
>     provideSharedPref(androidContext())
>   } withOptions {
>     named("SDKSharedPref")
>     createdAtStart()
>   }
> }
> 
> private fun provideSharedPref(
>   context: Context,
>   name: String = context.packageName
> ): SharedPreferences = context.getSharedPreferences(name, MODE_PRIVATE)
> 
> private val sdkModule = module {
>     single {
>       provideSDK(androidContext()) {
>         token = sharedPreferences.getString("token", "null") ?: "null"
>       }
>     } withOptions {
>       createdAtStart()
>     }
>   }
> ```

But when the provision is done using the annotation method, there is no longer a Missing Definition error.

```kotlin
@Module
internal class CoroutinesModule {
  @Singleton(createdAtStart = true)
  @Named("IO")
  fun provideIODispatcher() = Dispatchers.IO

  @Singleton(createdAtStart = true)
  @Named("Default")
  fun provideDefaultDispatcher() = Dispatchers.Default

  @Singleton(createdAtStart = true)
  @Named("Main")
  fun provideMainDispatcher() = Dispatchers.Main
}
```

**Comment by ghasemdev** [comment]
Perhaps a tag is needed to exclude the regular modules of koin. 
```kotlin
@IgnoreModule
internal val sharedPrefModule = module {
  single {
    provideSharedPref(androidContext())
  } withOptions {
    named("SDKSharedPref")
    createdAtStart()
  }
}
```

@arnaudgiuliani

**Comment by arnaudgiuliani** [comment]
Yeah, android elements must be ignored somewhere. It's a "known" issue of current implementation 👍

**Comment by arnaudgiuliani** [comment]
@ghasemdev interesting to also add somewhere @Ignore to avoid checking it if needed. Let's challenge it 🤔

**Comment by arnaudgiuliani** [comment]
Let's see #101
